mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+                        order,
+                        src_contract: None,
+                        src_decimals: None,
+                        src_amount: None,
+                        dst_contract: None,
+                        dst_amount: None,
+                        dst_decimals: None,
+                        chain_id: None,
+                        lr_price: None,
+                        lr_swap_data: None,
+                    })
+                })
+                .collect::<HashMap<(_, _), _>>(),
+        }
+    }
+
+    /// Calculate amounts of destination tokens required to fill ask orders for the requested base_amount:
+    /// multiplies base_amount by the order price (base_amount must be in 'coins')
+    async fn calc_destination_token_amounts(
+        &mut self,
+        ctx: &MmArc,
+        base_amount: &MmNumber,
+    ) -> MmResult<(), ApiIntegrationRpcError> {
+        for lr_data in self.inner.values_mut() {
+            let price: MmNumber = lr_data.order.price.rational.clone().into();
+            let dst_amount = base_amount * &price;
+            let coin = lp_coinfind_or_err(ctx, &lr_data.order.coin).await?;
+            lr_data.dst_amount = Some(wei_from_coins_mm_number(&dst_amount, coin.decimals())?);
+            log::debug!(
+                "calc_destination_token_amounts lr_data.order.coin={} coin.decimals()={} lr_data.dst_amount={:?}",
+                lr_data.order.coin,
+                coin.decimals(),
+                lr_data.dst_amount
+            );
+        }
+        Ok(())
+    }
+
+    fn update_with_lr_prices(&mut self, mut lr_prices: HashMap<(Ticker, Ticker), MmNumber>) {
+        for (key, val) in self.inner.iter_mut() {
+            val.lr_price = lr_prices.remove(key);
+        }
+    }
+
+    fn update_with_lr_swap_data(&mut self, mut lr_swap_data: HashMap<(Ticker, Ticker), ClassicSwapData>) {
+        for (key, val) in self.inner.iter_mut() {
+            val.lr_swap_data = lr_swap_data.remove(key);
+        }
+    }
+
+    async fn update_with_contracts(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        for ((src_token, dst_token), lr_data) in self.inner.iter_mut() {
+            let (src_coin, src_contract) = get_coin_for_one_inch(ctx, src_token).await?;
+            let (dst_coin, dst_contract) = get_coin_for_one_inch(ctx, dst_token).await?;
+            let src_decimals = src_coin.decimals();
+            let dst_decimals = dst_coin.decimals();
+            if src_decimals == 0 {
+                return MmError::err(ApiIntegrationRpcError::InternalError(format!(
+                    "Source token '{}' has invalid decimals (0)",
+                    src_token
+                )));
+            }
+            if dst_decimals == 0 {
+                return MmError::err(ApiIntegrationRpcError::InternalError(format!(
+                    "Destination token '{}' has invalid decimals (0)",
+                    dst_token
+                )));
+            }
```

Comment:
nit: could we make it so we can skip such an order gracefully?

Reply:
decimals is ensured to have a proper value at the coin activation and here we just re-check ourselves.
I suggest just using assert for tests instead (because decimals field is not checked when EthCoin created in tests)